File: src/main/java/com/chessonline/King.java
package com.chessonline;

import java.util.ArrayList;
import java.util.List;

public class King extends Piece {

    public King(String color, String position) {
        super("king", color, position);
    }

    @Override
    public boolean move(String newPosition) {

        if (!isMovementValid(newPosition)) return false;

        setPosition(newPosition);

        return true;
    }

    @Override
    public boolean isMovementValid(String newPosition) {
        Position current = new Position(position);
        Position target = new Position(newPosition);

        int rowDiff = Math.abs(target.row - current.row);
        int colDiff = Math.abs(target.col - current.col);

        if (rowDiff > 1 || colDiff > 1) {
            return false;
        }

        return true;
    }

    @Override
    public boolean isCaptureMovementValid(String newPosition) {
        return isMovementValid(newPosition);
    }

    public List<String> getPossibleMoves() {
        List<String> moves = new ArrayList<>();
        int[][] directions = {
            {1, 0}, {-1, 0}, {0, 1}, {0, -1},
            {1, 1}, {-1, -1}, {1, -1}, {-1, 1}
        };

        Position current = new Position(position);
        
        for (int[] dir : directions) {
            char newCol = (char) (current.col + dir[0]);
            int newRow = current.row + dir[1];

            if (isWithinBounds(newCol, newRow)) {
                moves.add("" + newCol + newRow);
            }
        }

        return moves;
    }

    private boolean isWithinBounds(char col, int row) {
        return col >= 'a' && col <= 'h' && row > 0 && row <= 8;
    }
}

File: engine/pieces/king.py
from .piece import Piece

class King(Piece):
    def __init__(self, color: str, position: str):
        super().__init__("king", color, position)

    def move(self, new_position: str) -> bool:
        if not self.is_movement_valid(new_position):
            return False

        self.set_position(new_position)
        return True

    def is_movement_valid(self, new_position: str) -> bool:
        current_col = self.position[0]
        current_row = int(self.position[1])

        target_col = new_position[0]
        target_row = int(new_position[1])

        row_diff = abs(target_row - current_row)
        col_diff = abs(ord(target_col) - ord(current_col))

        if row_diff > 1 or col_diff > 1:
            return False

        return True

    def is_capture_movement_valid(self, new_position: str) -> bool:
        return self.is_movement_valid(new_position)

    def get_possible_moves(self):
        moves = []

        directions = [
            (1, 0), (-1, 0), (0, 1), (0, -1),
            (1, 1), (-1, -1), (1, -1), (-1, 1)
        ]

        current_col = self.position[0]
        current_row = int(self.position[1])

        for dcol, drow in directions:
            new_col = chr(ord(current_col) + dcol)
            new_row = current_row + drow

            if self.is_within_bounds(new_col, new_row):
                moves.append(f"{new_col}{new_row}")

        return moves

    def is_within_bounds(self, col, row):
        return 'a' <= col <= 'h' and 1 <= row <= 8